### PR TITLE
Add e2e tests

### PR DESCRIPTION
**Pull request by mcbenjemaa**
This PR:

- includes capi test framework and e2e tests

Ref: https://release-1-5.cluster-api.sigs.k8s.io/developer/e2e

### CODE REVIEW

**Review by wikkyk — COMMENTED** [review]
Re commit 338e217db2605b5e83ab71d455ed227abbe85c3d
The licence header check is now removed pending a better replacement so you can make the copyright year 2024.

**Comment by wikkyk** [comment]
Depends on #75

File: go.mod
Comment (wikkyk) [inline go.mod:5]: do we actually want this?

File: test/e2e/capi_test.go
Comment (wikkyk) [inline test/e2e/capi_test.go:59]: sonarcloud is going to be unhappy :(

File: test/e2e/suite_test.go
Comment (wikkyk) [inline test/e2e/suite_test.go:235]: Can we take care of this now?

File: .github/workflows/e2e.yml
Comment (wikkyk) [inline .github/workflows/e2e.yml:22]: I think we'll need `pull_request_target` here. Otherwise this will always fail in PRs from forks due to the unavailable secrets.

File: go.mod
Comment (mcbenjemaa) [reply go.mod:5]: yes

Comment (mcbenjemaa) [reply go.mod:5]: This will make sure to replace the nested dependencies of CAPI to always v1.6.1

File: test/e2e/capi_test.go
Comment (mcbenjemaa) [reply test/e2e/capi_test.go:59]: sonarcloud is disabled in e2e

File: .github/workflows/e2e.yml
Comment (mcbenjemaa) [reply .github/workflows/e2e.yml:22]: Good point.

File: test/e2e/capi_test.go
Comment (wikkyk) [reply test/e2e/capi_test.go:59]: yes but it'll still count against the score :)

**Comment by wikkyk** [comment]
I rebased this for cluster-api 1.6.1, ref #18

**Comment by mcbenjemaa** [comment]
tests is gone

**Comment by avorima** [comment]
What about using github environments to control when and by whom these tests get run?
https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment

There still is the e2e environment from my initial testing that constantly is failing.

**Comment by mcbenjemaa** [comment]
> What about using github environments to control when and by whom these tests get run? https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment
> 
> There still is the e2e environment from my initial testing that constantly is failing.

yes, good point, 
I don't know how to run the fork there though 😄

**Comment by wikkyk** [comment]
We could but `pull_request_target` is nice and simple. @avorima we can use environments if you can come up with a working workflow but I don't want the PR to be blocked on this.

File: sonar-project.properties
Comment (wikkyk) [inline sonar-project.properties:5]: why?

Comment (mcbenjemaa) [reply sonar-project.properties:5]: this would include the testing into coverage, 
we don't want that.

Comment (wikkyk) [reply sonar-project.properties:5]: tests for tests! :)

**Comment by avorima** [comment]
> We could but `pull_request_target` is nice and simple. @avorima we can use environments if you can come up with a working workflow but I don't want the PR to be blocked on this.

The documentation for this event says: `Avoid using this event if you need to build or run code from the pull request.`. It even expands on this in the *Warning* section.
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target

To me that sounds like a security risk which should not be merged.

**Review by avorima — CHANGES_REQUESTED** [review]

File: go.mod
Comment (avorima) [reply go.mod:5]: Please explain why this is needed

Comment (mcbenjemaa) [reply go.mod:5]: There's other imported dependencies that have capi as dependency but different version, which leads to a conflict

File: Makefile
Comment (mcbenjemaa) [reply Makefile:281]: yes, 
but this one is to install the binary to be able to execute the tests

Comment (wikkyk) [reply Makefile]: You can get it from go.mod. See my comment on [f5033ae](https://github.com/ionos-cloud/cluster-api-provider-proxmox/pull/48/commits/f5033aeff164be2b18a5628b52f0ad8d964b4593)

File: Makefile
Comment (wikkyk) [inline Makefile:283]: It does assume there is only one instance of `github.com/onsi/ginkgo/v2 ` in go.mod, which should be the case.

File: Makefile
Comment (wikkyk) [reply Makefile]: The same goes for envsubst just below. We might want to use tools for this, instead.

File: go.mod
Comment (avorima) [reply go.mod:5]: Then please add a TODO comment for removing the replace and name the problematic dependencies

File: Makefile
Comment (avorima) [inline Makefile:294]: Just use `github.com/onsi/ginkgo/v2/ginkgo`. The version is fetched from go.mod if it's not provided

Comment (mcbenjemaa) [reply Makefile:294]: Wiktor already pushed that change.

Comment (wikkyk) [reply Makefile:294]: Good to know, so can we just delete the version checking dance altogether, then?

Comment (avorima) [reply Makefile:294]: If you're referring to the controller-gen and kustomize version checking, then not quite. Unless you reinstall everything on each make invocation, you must somehow check that the local binary has the correct version.
`go install github.com/onsi/ginkgo/v2/ginkgo` will just ensure that the version it uses matches go.mod. Update need to be handled separately.

Comment (avorima) [reply Makefile:294]: At least get rid of the `GINKGO_PKG` variable

File: test/e2e/README.md
Comment (avorima) [inline test/e2e/README.md:9]: This is provided by the Makefile, so it's not a requirement to run the tests

File: .github/workflows/e2e.yml
Comment (wikkyk) [reply .github/workflows/e2e.yml]: Why?

File: Makefile
Comment (wikkyk) [reply Makefile]: and replace it with what?

File: .github/workflows/e2e.yml
Comment (avorima) [reply .github/workflows/e2e.yml]: Just thinking about the future. Release branches are standard in the SIG, so this project should adopt them sooner or later.

Comment (wikkyk) [reply .github/workflows/e2e.yml]: We'll cross that bridge when we get there. Why shouldn't e2e tests run on all PRs?

**Comment by mcbenjemaa** [comment]
@wikkyk  did i mess it? 😄

File: .github/workflows/e2e.yml
Comment (avorima) [reply .github/workflows/e2e.yml]: I didn't think that we'd wanted to start using permanent feature branches. The normal process is to create a PR from main and merge into main. We have the same thing in place for non-e2e tests.

Comment (mcbenjemaa) [reply .github/workflows/e2e.yml]: Maybe if we could add a check that a maintainer should approve running the tests for specific PRs

Comment (avorima) [reply .github/workflows/e2e.yml]: I mean that's kind of what the environment is for. You can configure required reviews, which means that a maintainer must approve the run. Regardless, limiting this workflow to PRs for release branches doesn't seem controversial to me.

Comment (mcbenjemaa) [reply .github/workflows/e2e.yml]: I don't know what's required in this case.

File: go.mod
Comment (mcbenjemaa) [reply go.mod:5]: Note these circular dependencies are problematic for almost all providers.

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/go.mod#L5

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/go.mod#L5

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/go.mod#L8

https://github.com/kubernetes-sigs/cluster-api-provider-digitalocean/blob/main/go.mod#L132

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/go.mod#L222

https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/go.mod#L9


I don't know how to document it and which deps are causing this circular dependency.

This is a way in go module to pin a dependency version.

File: Makefile
Comment (mcbenjemaa) [reply Makefile]: @avorima will approve this

File: go.mod
Comment (wikkyk) [reply go.mod:5]: I agree. This should point at the version we're supporting.

**Review by wikkyk — APPROVED** [review]
ok with me if it passes selftests

**Comment by mcbenjemaa** [comment]
@avorima check the open points

**Comment by wikkyk** [comment]
blocks #92

**Review by wikkyk — APPROVED** [review]

**Review by avorima — APPROVED** [review]

**Pull request merged**
